src/node/Main.hs

```diff
@@ -0,0 +1,319 @@
+                            (CLI.expDistr commonArgs)
+#else
+                genesisStakeDistribution
+#endif
+        , npTimeLord = timeLord
+        , npJLFile = jlPath
+        , npAttackTypes = maliciousEmulationAttacks
+        , npAttackTargets = maliciousEmulationTargets
+        , npPropagation = not (CLI.disablePropagation commonArgs)
+        , npUpdatePath = updateLatestPath
+        , npUpdateWithPkg = updateWithPackage
+        , npUpdateServers = CLI.updateServers commonArgs
+        , npReportServers = CLI.reportServers commonArgs
+        }
+
+gtSscParams :: Args -> VssKeyPair -> GtParams
+gtSscParams Args {..} vssSK =
+    GtParams
+    { gtpSscEnabled = True
+    , gtpVssKeyPair = vssSK
+    }
+
+#ifdef WITH_WEB
+plugins :: (SscConstraint ssc, WorkMode ssc m) => Args -> [m ()]
+plugins Args {..}
+    | enableWeb = [serveWebBase webPort]
+    | otherwise = []
+#endif
+
+#ifdef WITH_WEB
+pluginsGT :: (WorkMode SscGodTossing m) => Args -> [m ()]
+pluginsGT Args {..}
+    | enableWeb = [serveWebGT webPort]
+    | otherwise = []
+#endif
+
+updateTriggerWorker
+    :: SscConstraint ssc
+    => ([WorkerSpec (RawRealMode ssc)], OutSpecs)
+updateTriggerWorker = first pure $ worker mempty $ \_ -> do
+    logInfo "Update trigger worker is locked"
+    void $ liftIO . takeMVar . ncUpdateSemaphore =<< getNodeContext
+    triggerShutdown
+
+walletServe
+    :: SscConstraint ssc
+    => Args
+    -> ([WorkerSpec (RawRealMode ssc)], OutSpecs)
+#if defined WITH_WEB && defined WITH_WALLET
+walletServe Args {..} =
+    if enableWallet
+    then first pure $ worker walletServerOuts $ \sendActions ->
+            walletServeWebFull sendActions walletDebug walletDbPath
+                                           walletRebuildDb walletPort
+    else updateTriggerWorker
+#else
+walletServe _ = updateTriggerWorker
+#endif
+
+walletProd
+    :: SscConstraint ssc
+    => Args
+    -> ([WorkerSpec (ProductionMode ssc)], OutSpecs)
+walletProd = first (map liftPlugin) . walletServe
+  where
+    liftPlugin (ActionSpec p) = ActionSpec $ \vI sa ->
+        lift . p vI $ hoistSendActions getNoStatsT lift sa
+
+walletStats
+    :: SscConstraint ssc
+    => Args
+    -> ([WorkerSpec (StatsMode ssc)], OutSpecs)
+walletStats = first (map liftPlugin) . walletServe
+  where
+    liftPlugin (ActionSpec p) = ActionSpec $ \vI sa -> do
+        s <- getStatsMap
+        lift . p vI $ hoistSendActions (runStatsT' s) lift sa
+
+printFlags :: IO ()
+printFlags = do
+#ifdef DEV_MODE
+    putText "[Attention] We are in DEV mode"
+#else
+    putText "[Attention] We are in PRODUCTION mode"
+#endif
+#ifdef WITH_WEB
+    putText "[Attention] Web-mode is on"
+#endif
+#ifdef WITH_WALLET
+    putText "[Attention] Wallet-mode is on"
+#endif
+    inAssertMode $ putText "Asserts are ON"
+
+main :: IO ()
+main = do
+    printFlags
+    args <- getNodeOptions
+    params <- baseParams "node" args
+    bracketResources params (action args)
```

Comment:
This `bracketResources` has type:

```haskell
bracketResources :: BaseParams -> (RealModeResources -> Production a) -> IO a
```

[Link to code in other review](https://github.com/fpco/cardano-sl/pull/3/files#diff-8cf34725c214de659667af4f8a642b7fR459).

I haven't tracked down which package the `Mockable.Production` comes from yet, but I don't understand why in dev *or* production mode, we run in a `Production` monad. Perhaps the "mockable" stuff isn't being used?